runtimevar/filevar/filevar.go

```diff
@@ -101,80 +107,111 @@ func (s *state) UpdateTime() time.Time {
+	// closeCh is used to return any errors from closing the notifier
+	// back to watcher.Close.
+	closeCh chan error
+	// shutdown tells the background goroutine to exit.
+	shutdown func()
 }
 
-// errorState returns a new State with err, unless prevS also represents
-// the same error, in which case it returns nil.
-func errorState(err error, prevS driver.State) driver.State {
-	s := &state{err: err}
-	if prevS == nil {
-		return s
+// WatchVariable implements driver.WatchVariable.
+func (w *watcher) WatchVariable(ctx context.Context, _ driver.State) (driver.State, time.Duration) {
+	select {
+	case <-ctx.Done():
+		return &state{err: ctx.Err()}, 0
+	case cur := <-w.ch:
+		return cur, 0
 	}
-	prev := prevS.(*state)
-	if prev.err == nil {
-		// New error.
+}
+
+// updateState checks to see if s and prev both represent the same error.
+// If not, it drains any previous state buffered in w.ch, then writes s to it.
+// It always return s.
+func (w *watcher) updateState(s, prev *state) *state {
+	if s.err != nil && prev != nil && prev.err != nil && (s.err == prev.err || (os.IsNotExist(s.err) && os.IsNotExist(prev.err))) {
+		// s represents the same error as prev.
 		return s
 	}
-	if err == prev.err {
-		return nil
-	}
-	if os.IsNotExist(err) && os.IsNotExist(prev.err) {
-		return nil
+	// Drain any buffered value on ch; it is now stale.
+	select {
+	case <-w.ch:
+	default:
 	}
+	// This write can't block, since we're the only writer, ch has a buffer
+	// size of 1, and we just read anything that was buffered.
+	w.ch <- s
 	return s
 }
 
-// WatchVariable implements driver.WatchVariable.
-func (w *watcher) WatchVariable(ctx context.Context, prev driver.State) (driver.State, time.Duration) {
-	// Start watching over the file and wait for events/errors.
-	// We can skip this if prev == nil, as we'll always immediately
-	// return a value in that case.
-	// We'll get a notifierErr if the file doesn't currently exist.
-	var notifierErr error
-	if prev != nil {
-		notifierErr = w.notifier.Add(w.file)
-		if notifierErr == nil {
-			defer func() {
-				_ = w.notifier.Remove(w.file)
-			}()
-		}
-	}
+// watch is run by a background goroutine.
+// It watches file using notifier, and writes new states to w.ch.
+// If it can't read or watch the file, it re-checks every waitTime.
+// It exits when ctx is canceled, and writes any shutdown errors (or
+// nil if there weren't any) to w.closeCh.
+func (w *watcher) watch(ctx context.Context, notifier *fsnotify.Watcher, file string, decoder *runtimevar.Decoder, waitTime time.Duration) {
+	// addedToNotifier is true iff file has been added to the notifier.
+	addedToNotifier := false
+	var cur *state
 
 	for {
+		// If the current state is an error, pause between attempts
+		// to avoid spin loops. In particular, this happens when the file
+		// doesn't exist.
+		if cur != nil && cur.err != nil {
+			select {
+			case <-ctx.Done():
+				w.closeCh <- notifier.Close()
+				return
+			case <-time.After(waitTime):
```

Comment:
time's godoc says "The underlying Timer is not recovered by the garbage collector until the timer fires." I think this is fine, it will be cleaned up shortly and there aren't going to be millions of these.